### PR TITLE
Add RHEL/Centos support to the system/users metricset

### LINKED ISSUE #16757: [Metricbeat] Support systemd version from RHEL 7 for system service metricset
RHEL 7 uses systemd version 219. Therefore the system service metricset does not work as the `ListUnitsByPatterns` (see https://github.com/elastic/beats/blob/d89675fc3bc4f6a8e72671ed887bf96ad8ca159b/metricbeat/module/system/service/service.go#L80)  method call did not exist in that version.

It seems that this method call was introduced in a later version of systemd:  https://github.com/systemd/systemd/commit/313fe66fbd0cd3d62566edece7944d6cf45b1b21

As an alternative `ListUnitsFiltered` could be used, as this method call is available since earlier versions of systemd and provides similar results to my understanding.

RHEL 7 is a widely used Linux system in enterprises which will get "normal" support until 30 June 2024.

### PR DESCRIPTION
**Pull request by fearful-symmetry**
## What does this PR do?

This PR is a simultaneous bugfix/enhancement that addresses both  #16757  and #16753. Rather than use the dbus `ListUnitsByPatterns` method, we make an introspection call to dbus, look at the methods available to us, and then return a function pointer to make a call dependent on what is available. The goal is to make systemd do as much of the work is possible, only falling back to filtering methods within metricbeat as needed. The filtering behavior implemented here is designed to emulate what happens on the dbus end via `ListUnitsByPatterns`. This also exposes `pattern_filter` since it was a 2-line change, and should be useful.

## Why is it important?

Currently, this metricset doesn't work on Centos7/RHEL7, due to the `ListUnitsByPatterns` method being too new.

## Checklist

- [X] My code follows the style guidelines of this project
- [X] I have commented my code, particularly in hard-to-understand areas
- [X] I have made corresponding changes to the documentation
- [X] I have made corresponding change to the default configuration files
- [X] I have added tests that prove my fix is effective or that my feature works

## Author's Checklist

- [x] Additonal testing needs to be done on centos and RHEL.

## How to test this PR locally

- Pull down and build the branch on centos 7.
- Run this metricset. Optionally, add some `pattern_filter` or `state_filter` values.

## Related issues

Closes elastic/beats#16757 and elastic/beats#16753

### CODE REVIEW

File: metricbeat/module/system/service/dbus.go [hunk 1]
```diff
@@ -0,0 +1,181 @@
...
+	//return a function callback ordered by desirability
+	if _, ok := unitMap["ListUnitsByPatterns"]; ok {
+		return listUnitsByPatternWrapper, nil
+	} else if _, ok := unitMap["ListUnitsFiltered"]; ok {
+		return listUnitsFilteredWrapper, nil
+	} else if _, ok := unitMap["ListUnits"]; ok {
+		return listUnitsWrapper, nil
+	}
+	return nil, fmt.Errorf("No supported list Units function: %v", unitMap)
+
```
Comment (jsoriano) [inline metricbeat/module/system/service/dbus.go]:
```suggestion
	return nil, fmt.Errorf("no supported list Units function: %v", unitMap)
```

File: metricbeat/module/system/service/dbus.go [hunk 2]
```diff
@@ -0,0 +1,181 @@
...
+	err = conn.Hello()
+	if err != nil {
+		return nil, errors.Wrap(err, "error in Hello")
+	}
+
+	var props string
+
+	//call "introspect" on the systemd1 path to see what ListUnit* methods are available
+	obj := conn.Object("org.freedesktop.systemd1", dbusRaw.ObjectPath("/org/freedesktop/systemd1"))
+	err = obj.Call("org.freedesktop.DBus.Introspectable.Introspect", 0).Store(&props)
```
Comment (jsoriano) [inline metricbeat/module/system/service/dbus.go]: Interesting!

**Comment by fearful-symmetry** [comment]
@jsoriano Testing this isn't particularly easy--we don't have any OS level integration tests, so I'm relying on testing parsers.

File: CHANGELOG.next.asciidoc [hunk 3]
```diff
@@ -112,6 +112,7 @@ https://github.com/elastic/beats/compare/v7.0.0-alpha2...master[Check the HEAD d
 - Fix detection and logging of some error cases with light modules. {pull}14706[14706]
 - Fix imports after PR was merged before rebase. {pull}16756[16756]
 - Add dashboard for `redisenterprise` module. {pull}16752[16752]
+- Dynamically choose a method for the system/service metricset to support older linux distros. {pull}16902[16902]
```
Comment (jsoriano) [inline CHANGELOG.next.asciidoc]: For a moment I thought that it should say `system/users` metricset, but is the title of the issue what is wrong. Remember this when forging the commit message! :slightly_smiling_face:

**Pull request merged**